# Esercizio Dal database AdventureWorks

### Estraiamo la tabella dimproduct

In [6]:
from sqlalchemy import create_engine
import pandas as pd

username = "studente_dapt"
password = "Ep1c0d3!!D4t4**4n4lys1s"
host = "epicode-data-pt-mysql.cvetyjye2qbl.eu-central-1.rds.amazonaws.com"
dbname = "AdventureWorksDW"

engine = create_engine(
    f"mysql+pymysql://{username}:{password}@{host}:3306/{dbname}"
)

df = pd.read_sql("SELECT * FROM dimproduct", engine)
print(df.head())


   ProductKey ProductAlternateKey  ProductSubcategoryKey  \
0           1             AR-5381                    NaN   
1           2             BA-8327                    NaN   
2           3             BE-2349                    NaN   
3           4             BE-2908                    NaN   
4           5             BL-2036                    NaN   

  WeightUnitMeasureCode SizeUnitMeasureCode     EnglishProductName  \
0                  None                None        Adjustable Race   
1                  None                None           Bearing Ball   
2                  None                None        BB Ball Bearing   
3                  None                None  Headset Ball Bearings   
4                  None                None                  Blade   

  SpanishProductName FrenchProductName  StandardCost  FinishedGoodsFlag  ...  \
0                                                NaN                  0  ...   
1                                                NaN      

## Sulla colonna DealerPrice, utilizzando il metodo .round(), arrotondiamo i valori alle due cifre decimali, e poi al valore intero più vicino

In [7]:
df["DealerPrice"] = df["DealerPrice"].round(2)
df["DealerPrice"] = df["DealerPrice"].round()
print(df)


     ProductKey ProductAlternateKey  ProductSubcategoryKey  \
0             1             AR-5381                    NaN   
1             2             BA-8327                    NaN   
2             3             BE-2349                    NaN   
3             4             BE-2908                    NaN   
4             5             BL-2036                    NaN   
..          ...                 ...                    ...   
601         602             BB-8107                    5.0   
602         603             BB-9108                    5.0   
603         604          BK-R19B-44                    2.0   
604         605          BK-R19B-48                    2.0   
605         606          BK-R19B-52                    2.0   

    WeightUnitMeasureCode SizeUnitMeasureCode     EnglishProductName  \
0                    None                None        Adjustable Race   
1                    None                None           Bearing Ball   
2                    None              

## Utilizzando il metodo .clip(), facciamo in modo che i valori siano compresi tra un minimo di 0 e un massimo di 1000

In [8]:
df["DealerPrice"] = df["DealerPrice"].clip(lower=0, upper= 55)
print(df)

     ProductKey ProductAlternateKey  ProductSubcategoryKey  \
0             1             AR-5381                    NaN   
1             2             BA-8327                    NaN   
2             3             BE-2349                    NaN   
3             4             BE-2908                    NaN   
4             5             BL-2036                    NaN   
..          ...                 ...                    ...   
601         602             BB-8107                    5.0   
602         603             BB-9108                    5.0   
603         604          BK-R19B-44                    2.0   
604         605          BK-R19B-48                    2.0   
605         606          BK-R19B-52                    2.0   

    WeightUnitMeasureCode SizeUnitMeasureCode     EnglishProductName  \
0                    None                None        Adjustable Race   
1                    None                None           Bearing Ball   
2                    None              

## Creiamo un DataFrame sintetico, che contiene i guadagni mensili di diverse annate, con il seguente codice:

years = 5
guadagni = pd.DataFrame({
    "Mese": list("GFMAMGLASOND") * years,   # 12 mesi ripetuti
    "Anno": np.repeat(list(range(1, years+1)), 12),  # 1..5
    "Valore_Guadagni_Mensili": np.random.randint(800, 5000, 12*years)
})

## Calcola la somma cumulativa dei guadagni utilizzando il metodo .cumsum()

In [14]:
import pandas as pd
import numpy as np

years = 5
guadagni = pd.DataFrame({
    "Mese": list("GFMAMGLASOND") * years,   # 12 mesi ripetuti
    "Anno": np.repeat(list(range(1, years+1)), 12),  # 1..5
    "Valore_Guadagni_Mensili": np.random.randint(800, 5000, 12*years)
})

guadagni["SumCumulativa"] = guadagni["Valore_Guadagni_Mensili"].cumsum()
print(guadagni)


   Mese  Anno  Valore_Guadagni_Mensili  SumCumulativa
0     G     1                     2513           2513
1     F     1                     1337           3850
2     M     1                     3020           6870
3     A     1                     4256          11126
4     M     1                     2722          13848
5     G     1                     2188          16036
6     L     1                     3698          19734
7     A     1                     1220          20954
8     S     1                     4454          25408
9     O     1                     1826          27234
10    N     1                     3498          30732
11    D     1                     3553          34285
12    G     2                     1123          35408
13    F     2                     4478          39886
14    M     2                     2556          42442
15    A     2                     1846          44288
16    M     2                     4731          49019
17    G     2               

## Come sopra, ma raggruppato per ogni anno usando prima un .groupby()

In [17]:
import pandas as pd
import numpy as np

years = 5
guadagni = pd.DataFrame({
    "Mese": list("GFMAMGLASOND") * years,   # 12 mesi ripetuti
    "Anno": np.repeat(list(range(1, years+1)), 12),  # 1..5
    "Valore_Guadagni_Mensili": np.random.randint(800, 5000, 12*years)
})

guadagni["SumCumulativa"] = guadagni.groupby("Anno")["Valore_Guadagni_Mensili"].cumsum()
print(guadagni)

   Mese  Anno  Valore_Guadagni_Mensili  SumCumulativa
0     G     1                     4329           4329
1     F     1                     2790           7119
2     M     1                     1560           8679
3     A     1                     1776          10455
4     M     1                     3873          14328
5     G     1                     3728          18056
6     L     1                     1051          19107
7     A     1                     2009          21116
8     S     1                     1280          22396
9     O     1                     3585          25981
10    N     1                     2937          28918
11    D     1                     3373          32291
12    G     2                     2581           2581
13    F     2                     3548           6129
14    M     2                     4671          10800
15    A     2                     3239          14039
16    M     2                     1536          15575
17    G     2               

## Dal database AdventureWorks estraiamo la tabella dimcustomer

In [19]:
df = pd.read_sql("SELECT * FROM dimcustomer", engine)
print(df.head())

   CustomerKey  GeographyKey CustomerAlternateKey Title  FirstName MiddleName  \
0        11000            26           AW00011000  None        Jon          V   
1        11001            37           AW00011001  None     Eugene          L   
2        11002            31           AW00011002  None      Ruben       None   
3        11003            11           AW00011003  None    Christy       None   
4        11004            19           AW00011004  None  Elizabeth       None   

  LastName  NameStyle   BirthDate MaritalStatus  ... EnglishOccupation  \
0     Yang          0  1971-10-06             M  ...      Professional   
1    Huang          0  1976-05-10             S  ...      Professional   
2   Torres          0  1971-02-09             M  ...      Professional   
3      Zhu          0  1973-08-14             S  ...      Professional   
4  Johnson          0  1979-08-05             S  ...      Professional   

  SpanishOccupation FrenchOccupation  HouseOwnerFlag  NumberCarsOwne

## Trasformiamo i nomi dei clienti in modo che abbiano solo lettere minuscole, e i cognomi in modo che abbiano solo lettere maiuscole

In [21]:
df["FirstName"] = df["FirstName"].str.lower
df["LastName"] = df["LastName"].str.upper

print(df)

       CustomerKey  GeographyKey CustomerAlternateKey Title  \
0            11000            26           AW00011000  None   
1            11001            37           AW00011001  None   
2            11002            31           AW00011002  None   
3            11003            11           AW00011003  None   
4            11004            19           AW00011004  None   
...            ...           ...                  ...   ...   
18479        29479           209           AW00029479  None   
18480        29480           248           AW00029480  None   
18481        29481           120           AW00029481  None   
18482        29482           179           AW00029482  None   
18483        29483           217           AW00029483  None   

                                               FirstName MiddleName  \
0      <bound method StringMethods.lower of <pandas.c...          V   
1      <bound method StringMethods.lower of <pandas.c...          L   
2      <bound method StringMet

## Sulla colonna EmailAddress, utilizzando il metodo .str.split(), estraiamo nome utente e dominio

In [27]:
df["EmailAddress"]
df["User"] = df["EmailAddress"].str.split("@").str[0]
df["Domain"] = df["EmailAddress"].str.split("@").str[1]

print(df[["User", "Domain"]])

             User               Domain
0           jon24  adventure-works.com
1        eugene10  adventure-works.com
2         ruben35  adventure-works.com
3       christy12  adventure-works.com
4      elizabeth5  adventure-works.com
...           ...                  ...
18479      tommy2  adventure-works.com
18480      nina21  adventure-works.com
18481       ivan0  adventure-works.com
18482    clayton0  adventure-works.com
18483      jésus9  adventure-works.com

[18484 rows x 2 columns]


## Sulla colonna Phone, estraiamo ogni parte del numero (ad es. da "1 (11) 500 555-0162" a ["1", "(11)", "500", "555-0162"])

In [ ]:
df["PhoneParts"] = df["Phone"].str.split(" ")
print(df)

       CustomerKey  GeographyKey CustomerAlternateKey Title  \
0            11000            26           AW00011000  None   
1            11001            37           AW00011001  None   
2            11002            31           AW00011002  None   
3            11003            11           AW00011003  None   
4            11004            19           AW00011004  None   
...            ...           ...                  ...   ...   
18479        29479           209           AW00029479  None   
18480        29480           248           AW00029480  None   
18481        29481           120           AW00029481  None   
18482        29482           179           AW00029482  None   
18483        29483           217           AW00029483  None   

                                               FirstName MiddleName  \
0      <bound method StringMethods.lower of <pandas.c...          V   
1      <bound method StringMethods.lower of <pandas.c...          L   
2      <bound method StringMet

## Utilizzando il metodo .str.contains(), estraiamo tutti gli indirizzi e-mail che contengono il numero "21"

In [31]:
df["EmailAddress"]
df.loc[df["EmailAddress"].str.contains("21", regex=False)]
print(df["EmailAddress"])

0             jon24@adventure-works.com
1          eugene10@adventure-works.com
2           ruben35@adventure-works.com
3         christy12@adventure-works.com
4        elizabeth5@adventure-works.com
                      ...              
18479        tommy2@adventure-works.com
18480        nina21@adventure-works.com
18481         ivan0@adventure-works.com
18482      clayton0@adventure-works.com
18483        jésus9@adventure-works.com
Name: EmailAddress, Length: 18484, dtype: object


## Estraiamo tutti gli indirizzi e-mail che contengono il numero "20" oppure il numero "10"

In [33]:
email_20_10 = df[df["EmailAddress"].str.contains("20|10")]
print(email_20_10)

       CustomerKey  GeographyKey CustomerAlternateKey Title  \
1            11001            37           AW00011001  None   
10           11010            22           AW00011010  None   
22           11022           609           AW00011022  None   
30           11030            28           AW00011030  None   
32           11032            35           AW00011032  None   
...            ...           ...                  ...   ...   
18395        29395           203           AW00029395  None   
18397        29397           193           AW00029397  None   
18415        29415           223           AW00029415  None   
18418        29418           163           AW00029418  None   
18428        29428           156           AW00029428  None   

                                               FirstName MiddleName  \
1      <bound method StringMethods.lower of <pandas.c...          L   
10     <bound method StringMethods.lower of <pandas.c...          C   
22     <bound method StringMet

## Calcolare la lunghezza di ogni indirizzo e-mail ed estrarre i cinque più lunghi e i cinque più corti

In [35]:
df["EmailLength"] = df["EmailAddress"].str.len()

df_long = df.sort_values("EmailLength", ascending=False).head(5)
df_short = df.sort_values("EmailLength", ascending=True).head(5)

print(df_long)
print(df_short)


       CustomerKey  GeographyKey CustomerAlternateKey Title  \
8768         19768           310           AW00019768  None   
979          11979            69           AW00011979  None   
14171        25171           543           AW00025171  None   
17017        28017           611           AW00028017  None   
1767         12767           648           AW00012767  None   

                                               FirstName MiddleName  \
8768   <bound method StringMethods.lower of <pandas.c...          C   
979    <bound method StringMethods.lower of <pandas.c...       None   
14171  <bound method StringMethods.lower of <pandas.c...          E   
17017  <bound method StringMethods.lower of <pandas.c...          B   
1767   <bound method StringMethods.lower of <pandas.c...          A   

                                                LastName  NameStyle  \
8768   <bound method StringMethods.upper of <pandas.c...          0   
979    <bound method StringMethods.upper of <pandas.

## Modificare il dominio degli indirizzi e-mail da "adventure-works.com" a "aw-db.com" mediante il metodo .str.replace()

In [36]:
df["NewEmailAddress"] = df["EmailAddress"].str.replace( "adventure-works.com", "aw-db.com")
print(df["NewEmailAddress"])

0             jon24@aw-db.com
1          eugene10@aw-db.com
2           ruben35@aw-db.com
3         christy12@aw-db.com
4        elizabeth5@aw-db.com
                 ...         
18479        tommy2@aw-db.com
18480        nina21@aw-db.com
18481         ivan0@aw-db.com
18482      clayton0@aw-db.com
18483        jésus9@aw-db.com
Name: NewEmailAddress, Length: 18484, dtype: object


## Dalla colonna AddressLine1 estraiamo tutti gli indirizzi che contengono la sottostringa "Street"

In [40]:
df_street = df[df["AddressLine1"].str.contains("Street", na=False)]
print(df_street["AddressLine1"])

7              942 Brook Street
12            4785 Scott Street
32           626 Bentley Street
63           4927 Virgil Street
125          8481 Zartop Street
                  ...          
18273        2155 Zartop Street
18292         3931 Scott Street
18323    134 Peachwillow Street
18333       1023 Hawkins Street
18351          8618 Rose Street
Name: AddressLine1, Length: 512, dtype: object


## Dai beginner_datasets carichiamo in un DataFrame il file facebook.csv, che contiene dei post con data di pubblicazione, tipo (foto, video, …) e numero di reactions raccolte:

In [13]:
import pandas as pd

df_fb = pd.read_csv("../beginner_datasets/facebook.csv")
print(df_fb.head())


                          status_id status_type status_published  \
0  246675545449582_1649696485147474       video   4/22/2018 6:00   
1  246675545449582_1649426988507757       photo  4/21/2018 22:45   
2  246675545449582_1648730588577397       video   4/21/2018 6:17   
3  246675545449582_1648576705259452       photo   4/21/2018 2:29   
4  246675545449582_1645700502213739       photo   4/18/2018 3:22   

   num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0            529           512         262        432         92         3   
1            150             0           0        150          0         0   
2            227           236          57        204         21         1   
3            111             0           0        111          0         0   
4            213             0           0        204          9         0   

   num_hahas  num_sads  num_angrys  
0          1         1           0  
1          0         0           0  
2          

## Con la funzione pd.to_datetime() convertiamo la colonna status_published in formato Timestamp

In [14]:
df_fb["Timestamp"] = pd.to_datetime(df_fb["status_published"])
print(df_fb.head())

                          status_id status_type status_published  \
0  246675545449582_1649696485147474       video   4/22/2018 6:00   
1  246675545449582_1649426988507757       photo  4/21/2018 22:45   
2  246675545449582_1648730588577397       video   4/21/2018 6:17   
3  246675545449582_1648576705259452       photo   4/21/2018 2:29   
4  246675545449582_1645700502213739       photo   4/18/2018 3:22   

   num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
0            529           512         262        432         92         3   
1            150             0           0        150          0         0   
2            227           236          57        204         21         1   
3            111             0           0        111          0         0   
4            213             0           0        204          9         0   

   num_hahas  num_sads  num_angrys           Timestamp  
0          1         1           0 2018-04-22 06:00:00  
1       

## Utilizzando gli attributi .dt.year , .dt.month , .dt.day , .dt.dayofweek , .dt.dayofyear, ottieniamo informazioni specifiche sulle date delle transazioni, come l'anno, il mese, il giorno della settimana, il giorno dell'anno, eccetera

In [15]:
df_fb["Timestamp"] = pd.to_datetime(df_fb["status_published"])
df_fb["Year"] = df_fb["Timestamp"].dt.year
df_fb["Month"] = df_fb["Timestamp"].dt.month
df_fb["Day"] = df_fb["Timestamp"].dt.day
df_fb["Day_of_Week"] = df_fb["Timestamp"].dt.dayofweek
print(df_fb[["Year", "Month", "Day", "Day_of_Week"]].head())

   Year  Month  Day  Day_of_Week
0  2018      4   22            6
1  2018      4   21            5
2  2018      4   21            5
3  2018      4   21            5
4  2018      4   18            2


## Estraiamo solo i post relativi al 2012

In [ ]:
df_2012 = df_fb[df_fb["Year"] == 2012]
print(df_2012.head())

Empty DataFrame
Columns: [status_id, status_type, status_published, num_reactions, num_comments, num_shares, num_likes, num_loves, num_wows, num_hahas, num_sads, num_angrys, Timestamp, Year, Month, Day, Day_of_Week]
Index: []


## Estraiamo solo i post relativi a maggio 2018

In [20]:
df_maggio_2018 = df_fb[(df_fb["Year"] == 2018) & (df_fb["Month"] == 5)]
print(df_maggio_2018.head())

                            status_id status_type status_published  \
2690  134115277150304_264554297439734       photo  5/31/2018 12:17   
2691  134115277150304_264493694112461       photo   5/31/2018 8:37   
2692  134115277150304_264416247453539       photo   5/31/2018 3:12   
2693  134115277150304_264168277478336       photo   5/30/2018 8:46   
2694  134115277150304_264160757479088       photo   5/30/2018 8:24   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2690             11             0           0         10          0         1   
2691             56             4           2         55          1         0   
2692             72            13           1         69          3         0   
2693             10             0           0         10          0         0   
2694             56             3          12         54          2         0   

      num_hahas  num_sads  num_angrys           Timestamp  Year  Month  Day  \
2690         

## Confrontiamo il numero di post pubblicati nei weekend rispetto al numero di post pubblicati nel resto della settimana

In [24]:
weekend_posts = df_fb[df_fb["Day_of_Week"].isin([5, 6])]
weekday_posts = df_fb[~df_fb["Day_of_Week"].isin([5, 6])]

print("Posts weekend:", len(weekend_posts))
print("Posts weekdays:", len(weekday_posts))


Posts weekend: 2041
Posts weekdays: 5009


## Troviamo il primo e ultimo post pubblicati in ogni anno

In [29]:
df_fb["Timestamp"] = pd.to_datetime(df_fb["status_published"])
df_fb["Year"] = df_fb["Timestamp"].dt.year

primi_ultimi = (df_fb.groupby("Year")["Timestamp"].agg(["min", "max"]))

print(primi_ultimi)


                     min                 max
Year                                        
2012 2012-07-15 02:51:00 2012-12-23 03:11:00
2013 2013-01-02 02:23:00 2013-12-31 23:38:00
2014 2014-01-05 09:23:00 2014-12-10 07:42:00
2015 2015-01-02 03:41:00 2015-12-31 08:20:00
2016 2016-01-03 04:23:00 2016-12-31 05:13:00
2017 2017-01-02 08:25:00 2017-12-31 22:15:00
2018 2018-01-01 01:39:00 2018-06-13 01:12:00


## Quanti tipi di post ci sono? E quanti per ogni tipo?

In [30]:
df_fb["status_type"].unique()
df_fb["status_type"].value_counts()

status_type
photo     4288
video     2334
status     365
link        63
Name: count, dtype: int64

## Dai beginner_dataset carichiamo in un DataFrame il file pokemon.csv:

In [33]:
import pandas as pd

df_poke = pd.read_csv("../beginner_datasets/pokemon.csv")
print(df_poke.head())

   #                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  


## Tramite i metodi .isnull() e .sum() controlliamo se ci sono valori nulli nel dataset e contiamo quanti valori nulli ci sono in ogni colonna

In [36]:
print(df_poke.isnull().sum().sum())

386


## Ci sono valori nulli?

Si, ce ne sono 386

## Se sì, avrebbe senso cercare di riempirli?

No, perchè non saprei che valore o stringa mettere. 

## Eliminiamo le righe che contengono valori nulli

In [42]:
df_poke = df_poke.dropna(subset=["Type 2"])
print(df_poke.isnull().sum().sum())

0


## Dai beginner_dataset carichiamo in un DataFrame il file automobile.csv:

In [44]:
import pandas as pd

df_auto = pd.read_csv("../beginner_datasets/automobile.csv")
print(df_auto.head())

   symboling  normalized-losses         make fuel-type aspiration  \
0          3                NaN  alfa-romero       gas        std   
1          3                NaN  alfa-romero       gas        std   
2          1                NaN  alfa-romero       gas        std   
3          2              164.0         audi       gas        std   
4          2              164.0         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0          130         mpfi  3.47    2.68       

## Ci sono valori nulli? Dove? Quanti?

In [45]:
print(df_auto.isnull().sum())

symboling             0
normalized-losses    37
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  0
stroke                0
compression-ratio     0
horsepower            0
peak-rpm              0
city-mpg              0
highway-mpg           0
price                 0
dtype: int64


normalized-losses    37
num-of-doors          2

## Quali righe hanno un valore nullo nella colonna num-of-doors?

In [46]:
df_auto[df_auto["num-of-doors"].isnull()]

symboling  normalized-losses   make fuel-type aspiration num-of-doors  \
26          1              148.0  dodge       gas      turbo          NaN   
60          0                NaN  mazda    diesel        std          NaN   

   body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
26      sedan          fwd           front        93.7  ...           98   
60      sedan          fwd           front        98.8  ...          122   

    fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
26         mpfi  3.03    3.39               7.6        102      5500       24   
60          idi  3.39    3.39              22.7         64      4650       36   

   highway-mpg  price  
26          30   8558  
60          42  10795  

[2 rows x 26 columns]

## Esaminando i dati nel dataset, cerchiamo una logica per sostituire i valori nulli nella colonna num-of-doors

In [49]:
df_auto["num-of-doors"].unique()

array(['two', 'four', nan], dtype=object)

In [50]:
print(df_auto[df_auto["body-style"] == "sedan"]["num-of-doors"].value_counts())

num-of-doors
four    79
two     14
Name: count, dtype: int64


In [55]:
df_auto.loc[df_auto["num-of-doors"].isnull(), "num-of-doors"] = "four"
df_auto

symboling  normalized-losses         make fuel-type aspiration  \
0            3                NaN  alfa-romero       gas        std   
1            3                NaN  alfa-romero       gas        std   
2            1                NaN  alfa-romero       gas        std   
3            2              164.0         audi       gas        std   
4            2              164.0         audi       gas        std   
..         ...                ...          ...       ...        ...   
197         -1               95.0        volvo       gas      turbo   
198         -1               95.0        volvo       gas        std   
199         -1               95.0        volvo    diesel      turbo   
200         -1               95.0        volvo       gas      turbo   
201         -1               95.0        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
197         four        sedan          rwd           front       109.1  ...   
198         four        sedan          rwd           front       109.1  ...   
199         four        sedan          rwd           front       109.1  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19     3.4              10.0        102   
4            136         mpfi  3.19     3.4               8.0        115   
..           ...          ...   ...     ...               ...        ...   
197          141         mpfi  3.78    3.15               8.7        160   
198          173         mpfi  3.58    2.87               8.8        134   
199          145          idi  3.01     3.4              23.0        106   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
197      5300       19          25  19045  
198      5500       18          23  21485  
199      4800       26          27  22470  
200      5400       19          25  22625  
201      5400       19          25  22625  

[202 rows x 26 columns]

## Abbiamo il seguente DataFrame che raccoglie le misurazioni di un sensore che misura la temperatura atmosferica giornaliera:

import numpy as np, pandas as pd temp = pd.DataFrame({"Giorno": [0, 1, 2, 3, 4, 5, 6 7, 8, 9, 10, 11, 12], "Temperature": [18, 19, 18, np.nan, 21, 20, 20, np.nan, 21, 23, np.nan, 23, 24]})

## Il sensore a volte non funziona, dunque alcuni dati sono mancanti: quale sarebbe la migliore strategia per gestirli?

In [58]:
import numpy as np
import pandas as pd

temp = pd.DataFrame({
    "Giorno": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    "Temperature": [18, 19, 18, np.nan, 21, 20, 20, np.nan, 21, 23, np.nan, 23, 24]
})

In [59]:
temp["Temperature"] = temp["Temperature"].interpolate()
print(temp)

    Giorno  Temperature
0        0         18.0
1        1         19.0
2        2         18.0
3        3         19.5
4        4         21.0
5        5         20.0
6        6         20.0
7        7         20.5
8        8         21.0
9        9         23.0
10      10         23.0
11      11         23.0
12      12         24.0


## Nel pacchetto os della standard library c'è la funzione os.listdir() che permette di avere la lista dei nomi di file all'interno di una directory; senza input di default li cerca nella directory di lavoro corrente, altrimenti si può passare un path per esaminare una directory specifica, ad esempio os.listdir("mio_progetto/beginner_datasets/")

## Nella directory dei beginner_datasets, quali sono i dataset che contengono dati nulli? 
Dovremo usare un ciclo for per esaminare tutti i nomi dei file

In [64]:
import os

path = "../beginner_datasets/"

for file in os.listdir(path):
    print(file)

amazon.csv
anomaly.csv
asia_gdp.csv
automobile.csv
bank.csv
bike.csv
blood.csv
boston.csv
cancer.csv
concrete.csv
credit.csv
diabetes.csv
diamond.csv
elections.csv
electrical_grid.csv
employee.csv
energy.csv
facebook.csv
forest.csv
france.csv
germany.csv
glass.csv
gold.csv
heart.csv
heart_disease.csv
hepatitis.csv
house.csv
income.csv
insurance.csv
ipl.csv
iris.csv
jewellery.csv
juice.csv
kiva.csv
mice.csv
migration.csv
nba.csv
parkinsons.csv
pokemon.csv
poker.csv
population.csv
public_health.csv
pycaret_datasets.xlsx
questions.csv
satellite.csv
seeds.csv
spx.csv
traffic.csv
tweets.csv
wholesale.csv
wikipedia.csv
wine.csv


Dovremo selezionare solo i nomi di file con estensione .csv (quindi usare un costrutto if)

In [ ]:
import os

path = "../beginner_datasets/"

for file in os.listdir(path):
    if file.endswith(".csv"): 
        print(file)

amazon.csv
anomaly.csv
asia_gdp.csv
automobile.csv
bank.csv
bike.csv
blood.csv
boston.csv
cancer.csv
concrete.csv
credit.csv
diabetes.csv
diamond.csv
elections.csv
electrical_grid.csv
employee.csv
energy.csv
facebook.csv
forest.csv
france.csv
germany.csv
glass.csv
gold.csv
heart.csv
heart_disease.csv
hepatitis.csv
house.csv
income.csv
insurance.csv
ipl.csv
iris.csv
jewellery.csv
juice.csv
kiva.csv
mice.csv
migration.csv
nba.csv
parkinsons.csv
pokemon.csv
poker.csv
population.csv
public_health.csv
questions.csv
satellite.csv
seeds.csv
spx.csv
traffic.csv
tweets.csv
wholesale.csv
wikipedia.csv
wine.csv


Nel corpo dovremo leggere di volta in volta il file in esame, e caricarlo in un DataFrame con la funzione .read_csv()

In [69]:
import os
import pandas as pd

path = "../beginner_datasets/"

for file in os.listdir(path):
    if file.endswith(".csv"):
        df = pd.read_csv(path + file)

Sul DataFrame dovremo utilizzare il metodo .isna() per trovare la maschera booleana dei dati nulli

In [70]:
import os
import pandas as pd

path = "../beginner_datasets/"

for file in os.listdir(path):
    if file.endswith(".csv"):
        df = pd.read_csv(path + file)
        mask_null = df.isna()

Dovremo contare i dati nulli, utilizzando .sum(); potremmo doverlo utilizzare più di una volta

In [71]:
import os
import pandas as pd

path = "../beginner_datasets/"

for file in os.listdir(path):
    if file.endswith(".csv"):
        df = pd.read_csv(path + file)
        mask_null = df.isna()

        null_per_col = mask_null.sum()
        total_null = null_per_col.sum()


Dovremo stampare, o memorizzare in una list, solo i nomi dei file che contengono dati nulli

In [72]:
import os
import pandas as pd

path = "../beginner_datasets/"
dataset_con_nulli = []

for file in os.listdir(path):
    if file.endswith(".csv"):
        df = pd.read_csv(path + file)
        mask_null = df.isna()
        total_null = mask_null.sum().sum()

        if total_null > 0:
            dataset_con_nulli.append(file)

print(dataset_con_nulli)

['automobile.csv', 'elections.csv', 'france.csv', 'hepatitis.csv', 'house.csv', 'income.csv', 'mice.csv', 'nba.csv', 'pokemon.csv', 'population.csv', 'seeds.csv', 'traffic.csv', 'wikipedia.csv']
